File: src/voxam/zmachine/dictionary.py
"""The dictionary table and lexical analysis (§13).

The dictionary lives in static memory at the address the header word
at $08 gives (§13.1): a list of word-separator codes, an entry length,
an entry count, and sorted entries whose first bytes are the encoded
text of each word (§13.2). Tokenization splits typed text at spaces
and separators; lookup encodes each word and hunts for its entry.
"""

from voxam.zmachine.memory import Memory
from voxam.zmachine.zscii import encode_word, zscii_to_char

# Entry text is 4 bytes through Version 3 and 6 bytes after (§13.3,
# §13.4).
V3_LAST_VERSION = 3
V3_TEXT_BYTES = 4
V4_TEXT_BYTES = 6

# An unrecognised word's dictionary address is 0 (§13.6.3).
NOT_IN_DICTIONARY = 0


class Dictionary:
    """A view of one dictionary table (§13.2).

    Usually the standard dictionary the header names, but tokenise
    may supply any table in the same format (§13.6).
    """

    def __init__(self, memory: Memory, base: int | None = None) -> None:
        """Read the table's header once (§13.2).

        Args:
            memory: The memory image holding the dictionary.
            base: The table's byte address; None means the standard
                dictionary named in the story header (§13.1).
        """

        self._memory = memory
        self._version = memory.header.version
        self._base = base if base is not None else memory.header.dictionary_address

        separator_count = memory.read_byte(self._base)

        self._separators = frozenset(
            zscii_to_char(memory.read_byte(self._base + 1 + index))
            for index in range(separator_count)
        )
        self._entry_length = memory.read_byte(self._base + 1 + separator_count)
        self._count = memory.read_word(self._base + 2 + separator_count)
        self._entries = self._base + 4 + separator_count
        self._text_bytes = (
            V3_TEXT_BYTES if self._version <= V3_LAST_VERSION else V4_TEXT_BYTES
        )

    @property
    def separators(self) -> frozenset[str]:
        """The word-separator characters (§13.2)."""

        return self._separators

    @property
    def entry_count(self) -> int:
        """How many words the dictionary holds (§13.2)."""

        return self._count

    def lookup(self, word: str) -> int:
        """Find a typed word's entry address, or 0 (§13.6.2).

        The word is encoded to dictionary form first, so lookup
        inherits the resolution guillotine: only the leading six or
        nine Z-characters distinguish words.

        Args:
            word: The typed word, however capitalized.

        Returns:
            The byte address of the entry, or 0 when absent.
        """

        target = encode_word(self._version, word)

        for index in range(self._count):
            address = self._entries + index * self._entry_length

            if self._text_at(address) == target:
                return address

        return NOT_IN_DICTIONARY

    def _text_at(self, address: int) -> bytes:
        """Read an entry's encoded text (§13.3)."""

        return bytes(
            self._memory.read_byte(address + offset)
            for offset in range(self._text_bytes)
        )


def tokenize(text: str, separators: frozenset[str]) -> list[tuple[str, int]]:
    """Split typed text into words (§13.6.1).

    Spaces divide words and are otherwise ignored; separators divide
    words while being words themselves.

    Args:
        text: The typed line.
        separators: The dictionary's word-separator characters.

    Returns:
        Each word with the offset it starts at within the text.
    """

    words: list[tuple[str, int]] = []
    start: int | None = None

    for position, character in enumerate(text):
        if character == " " or character in separators:
            if start is not None:
                words.append((text[start:position], start))
                start = None

            if character != " ":
                words.append((character, position))
        elif start is None:
            start = position

    if start is not None:
        words.append((text[start:], start))

    return words

File: csharp/Voxam.Core/Dictionary.cs
namespace Voxam.Core;

/// <summary>A dictionary table and lexical analysis (§13).</summary>
public sealed class DictionaryTable
{
    private readonly Memory _m;
    private readonly int _entryLength;
    private readonly int _count;
    private readonly int _entries;
    private readonly int _textBytes;

    public HashSet<char> Separators { get; } = [];

    public DictionaryTable(Memory m, int? baseAddress = null)
    {
        _m = m;
        var address = baseAddress ?? m.ReadWord(Header.Dictionary);
        var separatorCount = m.ReadByte(address);

        for (var i = 0; i < separatorCount; i++)
        {
            var text = Zscii.ToChar(m, m.ReadByte(address + 1 + i));

            if (text.Length == 1)
            {
                Separators.Add(text[0]);
            }
        }

        _entryLength = m.ReadByte(address + 1 + separatorCount);
        var count = m.ReadWord(address + 2 + separatorCount);

        if ((count & 0x8000) != 0)
        {
            count = 0x10000 - count;
        }

        _count = count;
        _entries = address + 4 + separatorCount;
        _textBytes = m.Version <= 3 ? 4 : 6;
    }

    /// <summary>A typed word's entry address, or 0 (§13.6.2).</summary>
    public int Lookup(string word)
    {
        var target = Zscii.EncodeWord(_m, word);

        for (var index = 0; index < _count; index++)
        {
            var address = _entries + index * _entryLength;
            var matched = true;

            for (var offset = 0; offset < _textBytes; offset++)
            {
                if (_m.ReadByte(address + offset) != target[offset])
                {
                    matched = false;
                    break;
                }
            }

            if (matched)
            {
                return address;
            }
        }

        return 0;
    }

    /// <summary>Split typed text into words with their offsets (§13.6.1).</summary>
    public static List<(string Word, int Offset)> Tokenize(string text, HashSet<char> separators)
    {
        var words = new List<(string, int)>();
        var start = -1;

        for (var position = 0; position < text.Length; position++)
        {
            var c = text[position];

            if (c == ' ' || separators.Contains(c))
            {
                if (start >= 0)
                {
                    words.Add((text[start..position], start));
                    start = -1;
                }

                if (c != ' ')
                {
                    words.Add((c.ToString(), position));
                }
            }
            else if (start < 0)
            {
                start = position;
            }
        }

        if (start >= 0)
        {
            words.Add((text[start..], start));
        }

        return words;
    }
}
